internal/config/incremental_sync.go

```diff
@@ -0,0 +1,207 @@
+package config
+
+import (
+	"context"
+	"github.com/icinga/icinga-go-library/database"
+	"github.com/icinga/icinga-go-library/types"
+	"github.com/jmoiron/sqlx"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"strings"
+)
+
+// IncrementalConfigurable specifies Getter methods required for types supporting incremental configuration loading.
+type IncrementalConfigurable[T comparable] interface {
+	zapcore.ObjectMarshaler
+
+	// GetID returns the primary key value.
+	GetID() T
+
+	// GetChangedAt returns the changed_at value.
+	GetChangedAt() types.UnixMilli
+
+	// IsDeleted returns if this entry was marked as deleted.
+	IsDeleted() bool
+
+	// Validate and optionally return an error, resulting in aborting this entry.
+	Validate() error
+}
+
+// buildSelectStmtWhereChangedAt creates a SQL SELECT for an incremental configuration synchronization.
+//
+// The query, which will be a prepared statement, expects a types.UnixMilli parameter to be compared against. This
+// parameter might be NULL, being COALESCEd to a numeric zero, returning all rows.
+//
+// The rows will be ordered by `changed_at`, allowing to update the last change timestamp when iterating over it.
+func (r *RuntimeConfig) buildSelectStmtWhereChangedAt(typePtr interface{}) string {
+	return r.db.Rebind(r.db.BuildSelectStmt(typePtr, typePtr) +
+		` WHERE "changed_at" > COALESCE(?, CAST(0 AS BIGINT)) ORDER BY "changed_at"`)
+}
+
+// selectRelationshipTableEntries constructs and execute a SELECT query for the config relationship tables.
+//
+// A SELECT query against the relationship tables is generated for either all entries at the initial sync (`changedAt`
+// has a NULL value) or a subset specified by either all `ids` in the `idField` or all changes since `changedAt`.
+//
+// The rows will be ordered by `changed_at`, allowing to update the last change timestamp when iterating over it.
+func (r *RuntimeConfig) selectRelationshipTableEntries(
+	ctx context.Context,
+	tx *sqlx.Tx,
+	typePtr interface{},
+	idField string,
+	ids []int64,
+	changedAt types.UnixMilli,
+	dest interface{},
+) error {
+	stmt := r.db.BuildSelectStmt(typePtr, typePtr) + ` WHERE "changed_at" > COALESCE(?, CAST(0 AS BIGINT))`
+	if len(ids) > 0 {
+		stmt += ` OR "` + idField + `" IN (` + strings.Join(strings.Split(strings.Repeat("?", len(ids)), ""), ",") + `)`
+	}
+	stmt += ` ORDER BY "changed_at"`
+	stmt = r.db.Rebind(stmt)
+
+	args := make([]interface{}, 0, 1+len(ids))
+	args = append(args, changedAt)
+	for id := range ids {
+		args = append(args, id)
+	}
```

Comment:
`sqlx.In()` should do most of this for you repeating placeholders and flattening/expanding the arguments.

Reply:
This function was dead code and could be removed.
See also https://github.com/Icinga/icinga-notifications/pull/191#discussion_r1613701885